crates/rpc/rpc-eth-api/src/helpers/call.rs

```diff
@@ -776,4 +779,430 @@ pub trait Call: LoadState + SpawnBlocking {
+                // when no fields are specified, set gas price to zero
+                let gas_price = gas_price.unwrap_or(U256::ZERO);
+                Ok(Self {
+                    gas_price,
+                    max_priority_fee_per_gas: None,
+                    max_fee_per_blob_gas: has_blob_hashes.then_some(block_blob_fee).flatten(),
+                })
+            }
+            (None, max_fee_per_gas, max_priority_fee_per_gas, None) => {
+                // request for eip-1559 transaction
+                let effective_gas_price = get_effective_gas_price(
+                    max_fee_per_gas,
+                    max_priority_fee_per_gas,
+                    block_base_fee,
+                )?;
+                let max_fee_per_blob_gas = has_blob_hashes.then_some(block_blob_fee).flatten();
+
+                Ok(Self {
+                    gas_price: effective_gas_price,
+                    max_priority_fee_per_gas,
+                    max_fee_per_blob_gas,
+                })
+            }
+            (None, max_fee_per_gas, max_priority_fee_per_gas, Some(max_fee_per_blob_gas)) => {
+                // request for eip-4844 transaction
+                let effective_gas_price = get_effective_gas_price(
+                    max_fee_per_gas,
+                    max_priority_fee_per_gas,
+                    block_base_fee,
+                )?;
+                // Ensure blob_hashes are present
+                if !has_blob_hashes {
+                    // Blob transaction but no blob hashes
+                    return Err(RpcInvalidTransactionError::BlobTransactionMissingBlobHashes.into())
+                }
+
+                Ok(Self {
+                    gas_price: effective_gas_price,
+                    max_priority_fee_per_gas,
+                    max_fee_per_blob_gas: Some(max_fee_per_blob_gas),
+                })
+            }
+            _ => {
+                // this fallback covers incompatible combinations of fields
+                Err(EthApiError::ConflictingFeeFieldsInRequest)
+            }
+        }
+    }
+}
+
+/// Applies the given block overrides to the env
+fn apply_block_overrides(overrides: BlockOverrides, env: &mut BlockEnv) {
+    let BlockOverrides {
+        number,
+        difficulty,
+        time,
+        gas_limit,
+        coinbase,
+        random,
+        base_fee,
+        block_hash: _,
+    } = overrides;
+
+    if let Some(number) = number {
+        env.number = number;
+    }
+    if let Some(difficulty) = difficulty {
+        env.difficulty = difficulty;
+    }
+    if let Some(time) = time {
+        env.timestamp = U256::from(time);
+    }
+    if let Some(gas_limit) = gas_limit {
+        env.gas_limit = U256::from(gas_limit);
+    }
+    if let Some(coinbase) = coinbase {
+        env.coinbase = coinbase;
+    }
+    if let Some(random) = random {
+        env.prevrandao = Some(random);
+    }
+    if let Some(base_fee) = base_fee {
+        env.basefee = base_fee;
+    }
+}
```

Comment:
move this to `reth-rpc-eth-types` pls

Reply:
also the tests!

Reply:
> move this to `reth-rpc-eth-types` pls

I move struct CallFees from path crates/rpc/rpc-eth-types/src/revm_utils.rs to crates/rpc/rpc-eth-api/src/helpers/call.rs.  function create_txn_env need to use CallFees. I need to make struct CallFees members all public. Is it ok? @emhane

Reply:
Fixed. Thanks.

Reply:
that's ok :) public is always preferable over private since reth libraries are built to be extensible. we don't always know how users may want to use parts of the library in new ways.

here, it's clearly useful for the type to be public, since we encourage different networks to override any default trait method implementations to suit their network, and they may need the `CallFees` type to do that. it's a waste if they have to redefine the type in their code because the type is private.

Reply:
got it.